File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course2_Data_Structures/week1_Basic_Data_Structures/check_brackets_in_code/check_brackets.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.io.IOException;
import java.util.Stack;

class Bracket {
    Bracket(char type, int position) {
        this.type = type;
        this.position = position;
    }

    boolean Match(char c) {
        if (this.type == '[' && c == ']')
            return true;
        if (this.type == '{' && c == '}')
            return true;
        if (this.type == '(' && c == ')')
            return true;
        return false;
    }

    char type;
    int position;
}

class check_brackets {
    public static void main(String[] args) throws IOException {
        InputStreamReader input_stream = new InputStreamReader(System.in);
        BufferedReader reader = new BufferedReader(input_stream);
        String text = reader.readLine();

        int ans = -1;
        Stack<Bracket> opening_brackets_stack = new Stack<Bracket>();
        for (int position = 0; position < text.length(); ++position) {
            char next = text.charAt(position);

            if (next == '(' || next == '[' || next == '{') {
                // Process opening bracket, write your code here
                opening_brackets_stack.push(new Bracket(next,position));
            }

            if (next == ')' || next == ']' || next == '}') {
                // Process closing bracket, write your code here
                if(opening_brackets_stack.isEmpty()||!opening_brackets_stack.peek().Match(next)){
                    ans = position +1;
                    break;
                }
                opening_brackets_stack.pop();
            }
        }

        // Printing answer, write your code here
        if(ans == -1 && opening_brackets_stack.isEmpty())
            System.out.println("Success");
        else if(ans == -1)
            System.out.println(opening_brackets_stack.peek().position+1);
        else
            System.out.println(ans);
    }
}

File: Specialization_Data_Structures_Algorithms_UC_San_Diego/Course2_Data_Structures/week1_Basic_Data_Structures/check_brackets_in_code/check_brackets.py
# python3

import sys

class Bracket:
    def __init__(self, bracket_type, position):
        self.bracket_type = bracket_type
        self.position = position

    def Match(self, c):
        if self.bracket_type == '[' and c == ']':
            return True
        if self.bracket_type == '{' and c == '}':
            return True
        if self.bracket_type == '(' and c == ')':
            return True
        return False

if __name__ == "__main__":
    text = sys.stdin.read()
    ans = -1
    opening_brackets_stack = []
    for i, next in enumerate(text):
        if next == '(' or next == '[' or next == '{':
            # Process opening bracket, write your code here
            opening_brackets_stack.append(Bracket(next,i))

        if next == ')' or next == ']' or next == '}':
            # Process closing bracket, write your code here
            if(len(opening_brackets_stack)==0 or not opening_brackets_stack[-1].Match(next)):
                ans = i +1
                break
            opening_brackets_stack.pop()
    # Printing answer, write your code here
    if(ans == -1 and len(opening_brackets_stack)==0):
        print("Success")
    elif(ans == -1):
        print(opening_brackets_stack[-1].position+1)
    else:
        print(ans)